lib/temporal/worker/reactor.rb

```diff
@@ -0,0 +1,112 @@
+require 'fiber'
+require 'forwardable'
+require 'set'
+require 'singleton'
+require 'temporal/worker/reactor/promise'
+require 'temporal/worker/reactor/command'
+
+module Temporal
+  class Worker
+    class Reactor
+      include Singleton
+      extend SingleForwardable
+
+      def_delegators :instance, :execute, :attach
+
+      def initialize
+        @queue = Queue.new
+        @fibers = Set.new
+        @thread = nil
+        @running = false
+        @mutex = Mutex.new
+      end
+
+      def execute(&block)
+        mutex.synchronize do
+          unless running?
+            @thread = Thread.new { run_loop }
```

Comment:
Does this have to be done in a thread? Can't it be a fiber that yields until work is available? Is there a fiber-capable queue?

Reply:
The reactor has to run in a thread otherwise it'll take over the current thread. If we replace it with a Fiber we'll need to somehow keep resuming it, but we're in the user-thread at that point and can't have anything running in it

Reply:
This may make too many assumptions about how a user may want to run things, but can leave for now.

Reply:
They can, but precisely because we don't know how a user is running things we need to shield ourselves from it and Thread is probably the most suitable container here